perf/impl/go-libp2p/v0.27/main.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	tls "github.com/libp2p/go-libp2p/p2p/security/tls"
+	"github.com/multiformats/go-multiaddr"
+)
+
+func main() {
+	runServer := flag.Bool("run-server", false, "Should run as server")
+	serverAddr := flag.String("server-address", "", "Server address")
+	transport := flag.String("transport", "tcp", "Transport to use")
+	secretKeySeed := flag.Uint64("secret-key-seed", 0, "Server secret key seed")
+	uploadBytes := flag.Uint64("upload-bytes", 0, "Upload bytes")
+	downloadBytes := flag.Uint64("download-bytes", 0, "Download bytes")
+	flag.Parse()
+
+	opts := []libp2p.Option{
+		// Use TLS only instead of both TLS and Noise. Removes the
+		// additional multistream-select security protocol negotiation.
+		// Thus makes it easier to compare with TCP+TLS+HTTP/2
+		libp2p.Security(tls.ID, tls.New),
+	}
+
+	if *runServer {
+		opts = append(opts, libp2p.ListenAddrStrings("/ip4/0.0.0.0/tcp/4001", "/ip4/0.0.0.0/udp/4001/quic-v1"))
+
+		// TODO: Fake identity. For testing only.
+		priv, _, err := crypto.GenerateEd25519Key(&simpleReader{seed: uint8(*secretKeySeed)})
+		if err != nil {
+			panic(err)
+		}
+		opts = append(opts, libp2p.Identity(priv))
+	}
+
+	h, err := libp2p.New(opts...)
+	if err != nil {
+		panic(err)
+	}
+
+	perf := NewPerfService(h)
+	if *runServer {
+		for _, a := range h.Addrs() {
+			fmt.Println(a.Encapsulate(multiaddr.StringCast("/p2p/" + h.ID().String())))
+		}
+
+		select {} // run forever, exit on interrupt
+	}
+
+	ipPort := strings.Split(*serverAddr, ":")
+	if len(ipPort) != 2 {
+		fmt.Println("Invalid server address format. Expected format: 'ip:port'")
+		return
+	}
+
+	ip := ipPort[0]
+	port := ipPort[1]
+
+	var multiAddrStr string
+	switch *transport {
+	case "tcp":
+		multiAddrStr = fmt.Sprintf("/ip4/%s/tcp/%s/p2p/12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN", ip, port)
+	case "quic-v1":
+		multiAddrStr = fmt.Sprintf("/ip4/%s/udp/%s/quic-v1/p2p/12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN", ip, port)
+	default:
+		fmt.Println("Invalid transport. Accepted values: 'tcp' or 'quic-v1'")
+		return
+	}
+
+	serverInfo, err := peer.AddrInfoFromString(multiAddrStr)
+	if err != nil {
+		panic(err)
+	}
+
+	start := time.Now()
+	h.Peerstore().AddAddrs(serverInfo.ID, serverInfo.Addrs, peerstore.TempAddrTTL)
+	// Use h.Network().DialPeer() instead of h.Connect to skip waiting for
+	// identify protocol to finish.
+	_, err = h.Network().DialPeer(context.Background(), serverInfo.ID)
+	if err != nil {
+		panic(err)
+	}
+	connectionEstablished := time.Since(start)
```

Comment:
Skipping the identify wait in `Connect`.

Reply:
@marten-seemann @MarcoPolo @sukunrt any idea why the connection establishment with the code above takes 3 network round-trips? I would expect it to take 2 RTT.

My assumption:
- 1 RTT for TCP.
- I only enable TLS, thus no security protocol negotiation, more specifically optimistic 0RTT multistream-select.
- 1 RTT for TLS. (Assuming TLS 1.3 is the default.)
- Using default muxers, which should be only Yamux, thus no muxer negotiation, more specifically optimistic 0RTT multistream-select.
- Not waiting for the additional identify exchange.

Reply:
go-libp2p will spend 1RTT negotiating the security. 
go-libp2p doesn't do 0RTT select for security. It does 0RTT select for protocols when it knows the peer supports the given protocol(mostly via identify) otherwise it'll still do the 1RTT select.

Reply:
> go-libp2p will spend 1RTT negotiating the security protocol.

Given that the client only supports one security protocol, i.e. TLS, shouldn't it be using the lazy mechanism, in other words optimistically use TLS @sukunrt?

https://github.com/multiformats/go-multistream/blob/master/lazyClient.go

Reply:
I think we can change this. But currently this is not what's happening. 
https://github.com/libp2p/go-libp2p/blob/master/p2p/net/upgrader/upgrader.go#L319-L330
In security negotiation we don't use lazyClient. 
It uses this function which does wait for the reply. https://github.com/multiformats/go-multistream/blob/master/client.go#L117

Reply:
Thank you @sukunrt. I forgot that go-libp2p supports [the multistream-select simultaneous-open extension](https://github.com/libp2p/specs/blob/master/connections/simopen.md) and thus can not do an optimistic multistream-select negotiation when negotiating a single protocol (here TLS) only.

The extension is not needed in libp2p's hole punching as roles (i.e. dialer and listener) are derived from the roles on the relayed connection, see [DCUtR protocol section](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol). As far as I can tell the simultaneous-open extension is only needed for the case where two nodes randomly connect via TCP at the same time.

In my eyes the probability for two nodes to randomly, i.e. not coordinated via DCUtR, connect via TCP **at the same time** is low and only leads to a connection failure. Thus we don't support it in rust-libp2p.

@sukunrt @MarcoPolo @marten-seemann what do you think of removing the simultaneous-open extension in go-libp2p and thus enabling the optimistic security protocol negotiation when negotiation a single protocol only? This would allow go-libp2p to be able to compete with vanilla TCP+TLS+HTTP when it comes to connection establishment latency. In other words both go-lib2p with TLS+Yamux and vanilla TCP+TLS+HTTP would take 3 RTT (2 RTT connection establishment + 1 RTT request/response).

Reply:
I don’t think we should make code changes just to look better in benchmarks. In practice, most configurations (IPFS and Lotus) use TLS and Noise.

@vyzo had numbers on simultaneous opens happening in the wild. Would you mind sharing them here?

Reply:
FWIW, in nim-libp2p, we don't bind when we are public, which avoids uncoordinated simultaneous opens in the wild

We don't have enough data yet to see if it causes issues, but so far the only scenario where it doesn't work is if someone made a manual port mapping to a different port than the internal one, and didn't specify it to libp2p. In this case, we can't discover our public port, and we think we are private (since we don't bind)

Reply:
Thank you for the input everyone.

I documented the above as well as past discussions in https://github.com/libp2p/go-libp2p/issues/2330. Let's continue the discussion over there in case there is interest.